### ISSUE TITLE
Memory use of strings

### ISSUE THREAD

**Issue by bitcraft** [Tuxemon/Tuxemon#288]
This is a suggestion to clean up the code base a little bit.  We've made a great effort overhauling the UI to support multiple languages, and now I'd like to see it taken a step further.

Many game strings live in memory as duplicates of other strings.  For example, every "potion" will contain the string "Potion", "Gives 50hp", etc.  Obviously not a huge memory loss, but it is duplicated, and will also prevent changing the game language on the fly (possibly also causing save game issues).

So....  here is my proposal!

- Make more use of the translator!  It should be the only way to get game strings.
- Remove all cached copies of game strings
- Rewrite game code which references cached game strings, and replace with trans object

For example, consider the following in the Item class:

```python
    def load(self, slug):
        """Loads and sets this items's attributes from the item.db database. The item is looked up
        in the database by slug.

        :param slug: The item slug to look up in the monster.item database.

        :type slug: String

        :rtype: None
        :returns: None
        """

        results = items.lookup(slug, table="item")
        self.slug = results["slug"]  # short English identifier
        self.name = trans(results["name_trans"])  # will be locale string
        self.description = trans(results["description_trans"])  # will be locale string

        self.type = results["type"]
        self.power = results["power"]
        self.sprite = results["sprite"]
        self.usable_in = results["usable_in"]
        self.target = db.process_targets(results["target"])
        self.effect = results["effects"]
        self.surface = tools.load_and_scale(self.sprite)
        self.surface_size_original = self.surface.get_size()
```

The slug can stay, but the name and description should stay as the untranslated string.  When items are referenced in the game, it will be translated each time.

If you wanted to get really crazy:
- Make a ItemClassFactory Metaclass
- For each item or technique leaded from the DB, create a new class
- make the "static" elements like animation, name, description class attributes
- New instances of a particular item will be their own class
- for example "PotionClass", "WhirlwindClass"

Benefits are reduced memory use for free, as all instances of a particular things will be sharing one set of data.

Anyway, I'll keep this issue open for discussion.  The 1st option I listed is doable, and should be done an 2 hours or so, after testing.  the 2nd option will require more effort, but it would be nice if implemented.

Thoughts?  Am I crazy?  Any takers?

**Comment by ShadowApex** [comment]
I don't think that the memory savings will be very significant. Even with a million items, the extra strings would take maybe a MB or two of extra memory. But I think the biggest win in doing it the way you describe is that save games will work perfectly, even if you switch locale mid-game. I think this is definitely something we should do.

**Comment by bitcraft** [comment]
I doubt the memory saving of strings would be significant either.  Images
and animations need to be de-duplicated in memory as well (do they?).
Something to think about.

The save game issue is important though.

**Comment by JoniHintz** [comment]
That sounds like something I can do. 
I would implement it using property decorators like in this example: https://www.journaldev.com/14893/python-property-decorator The advantage would be that no code outside of the class would have to be changed for this to work.

**Comment by bitcraft** [comment]
Hi @JoniHintz.  Could you provide an example?  My original though was to make a class factory for each DB entry, with all the default attributes as class attributes rather than instance attributes, then enforce items size using `__slots__`.  Kinda like a poor-man's flyweight pattern.  anyway, im interested in seeing your proposal.

**Comment by JoniHintz** [comment]
Sure, I replace each instance variable with a function of the same name and the property decorator. Inside I call the functions to get the translated string. 

Example:

`@property`
def use_item(self):
&nbsp;&nbsp;&nbsp;&nbsp; if self.slug: 
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;results = db.lookup(self.slug, table="item")
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return T.translate(results["use_item"])
&nbsp;&nbsp;&nbsp;&nbsp;else:
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;return ""

Outside of the class the property can be accessed like a normal instance variable.

This should work in items.py and technique.py and the read only parts of monster.py

**Comment by bitcraft** [comment]
`use_item` is a function...the caller should expect side-effects.  it doesn't make sense to make it a property; properties are a way to build getters and setters.
as for the the db lookups and translate calls, i don't know if thats the right approach.  it would reduce memory....but id rather have our translated strings in memory.  i might change my mind though.

**Comment by JoniHintz** [comment]
> if i am reading it correctly though, you are proposing making a property for every message that we would expect to use from the translations?  this is not maintainable because it means script writers will need to write python code to write the story.

 It shouldn't change the logic much when I replaces the strings with the property functions inside the class. Except (1) the translated string is looked up whenever it is accessed and (2) the string can not be simply set to another value without changing  slug. (1) Should take care of a locale change mid game.  I assumed that (2) wouldn't be an issue since setting it manually would circumvent translation. But now I am not sure. I might need to look into how the story is implemented to understand your concern.

> as for the the db lookups and translate calls, i don't know if thats the right approach.

Yup, I admit that is  hacky.  
An alternative could be to implement it flyweight style + an observer that  only looks the translation up when locale changes. But that seems a bit more complex than my first idea.

**Comment by bitcraft** [comment]
i do like the prospect of changing locale mid game.  at the same time, its not a huge concern for anyone at the moment.  its easy enough to save & restart the game.  i just feel like right now that the property approach is really boilerplate heavy, and anti-dry.  it looks like 7 lines of very similar code for each translated string that we expect an object would be using.

i agree that the flyweight pattern would apply here.  tbh, sweeping code base changes are not really a huge issue.  if the best way involves lot of changes, then that is ok as long as the code is not horrible.  for instance, if you decide that object attributes are too problematic and we should have a function called "get_translated_attribute" (for example), that could be considered.

i will mention though that the duplication of game sprites will not be a problem in the future, as i am in the process of writing a new rendering system that maintains its own sprites....the sprites on items/monsters/npcs etc will be removed and managed by another class.

i think that i wasn't correct when stating "because it means script writers will need to write python code to write the story.".  translated strings for game objects would always require a code change anyway, since we expect the string to be an attribute of the object.  so, sorry about the confusion there; my mistake.

**Comment by JoniHintz** [comment]
I am  convinced that using a function + property decorator is not the way to go, especially when changing locale mid game is no a priority right now.

While searching for best practices for implementing the flyweight-pattern with strings, I found the intern() function.
https://python-reference.readthedocs.io/en/latest/docs/functions/intern.html

An allocation would look like this:
` self.name = intern(trans(results["name_trans"]))`

**Comment by bitcraft** [comment]
Yeah i saw that too.  I think that it would be a quick win and worth trying out.

**Comment by JoniHintz** [comment]
I tried intern it in Item. The result was pretty underwhelming. 

If I am not mistaken Items are stateless, so this should be possible:
items = dict()
```
def get_item(game,user,item_slug):
    if item_slug in Item.items:
        return Item.items[item_slug]
    else:
        item.items[item_slug] = Item(game,user,item_slug)
        return Item.items[item_slug] 
```
This reduces the number of item instances noticeably.
The problem  (besides a minor overhead) is that it would be necessary to replace Item(... with Item.get_item(... 

I only found one occurrence of Item(... in the npc class for when a new item is added to the players inventory, but items are also created when a save is loaded or every time a shop is entered. I am currently looking into whether the function could be used there.
